### ISSUE TITLE
Make it easier to run mypy locally

### ISSUE THREAD

**Issue by samestep** [pytorch/pytorch#50513]
We currently have [wiki instructions][lint as you type] for running Flake8 locally in a couple editors, including VS Code. This makes it convenient to correct Flake8 errors before pushing, but we currently don't have similar documentation for running mypy locally. Given [how much effort][typing pulls] we've been putting into adding mypy annotations to the codebase, we should make this easier to do.

### Running all mypy checks

Currently the canonical way to run mypy on PyTorch is via the `test/test_type_hints.py` file; for example:
```
$ python test/test_type_hints.py --save-xml=/tmp/mypy_time
```
On each of the dev machines I have access to, the above command (as of 5834438090a1b3206347e30968e48f44251a53a1) takes about 4 minutes, the vast majority of which is for the `test_run_mypy` test (the one that uses `mypy.ini`, which applies to most of the Python files we typecheck):
```
$ test/print_test_stats.py /tmp/mypy_time | tail -n6
Total runtime is 0:04:07
4 longest tests of entire run:
    TestTypeHints.test_run_mypy  time: 165.97 seconds
    TestTypeHints.test_type_hint_examples  time: 40.09 seconds
    TestTypeHints.test_doc_examples  time: 29.65 seconds
    TestTypeHints.test_run_mypy_strict  time: 11.50 seconds
```
This is too long to include in a local development feedback loop. Note that even though [mypy uses a cache by default][mypy cache], running our mypy test suite takes the same amount of time each run, because it runs mypy using 4 different configs (one for each test case) and thus invalidates `.mypy_cache` every time.

### Running most mypy checks

If you're only editing files that use `mypy.ini` rather than `mypy-strict.ini` (which is most files in this repo), you can just run a single test from that test suite:
```
$ python test/test_type_hints.py TestTypeHints.test_run_mypy
```
After an initial warm-up run, this should only take a few seconds to complete. The runtime can go up significantly depending on what changes you've made since the last run; for instance, changing the return type of [`torch.random.set_rng_state`][] from `None` to `str` increases the time to about 90 seconds:
```diff
diff --git a/torch/random.py b/torch/random.py
index 31e2643845..7d92aafb40 100644
--- a/torch/random.py
+++ b/torch/random.py
@@ -5,7 +5,7 @@ from torch._C import default_generator
 import torch
 
 
-def set_rng_state(new_state) -> None:
+def set_rng_state(new_state) -> str:
     r"""Sets the random number generator state.
 
     Args:
```

### Running most mypy checks in VS Code

It is possible to enable mypy in VS Code by setting these values in `.vscode/settings.json`:
```json
{
  "python.linting.enabled": true,
  "python.linting.mypyEnabled": true
}
```
This has the advantage of displaying the warnings inline after every edit rather than having to repeatedly re-run a command, and anecdotally seems to run a bit faster than running the command, although it can still be slow. To use the strict profile, add this setting as well:
```json
  "python.linting.mypyArgs": [
    "--ignore-missing-imports",
    "--follow-imports=silent",
    "--show-column-numbers",
    "--config-file=/path/to/pytorch/mypy-strict.ini"
  ]
```
(The first three args are just the default VS Code mypy args.)

However, this setup has a couple problems (listed in decreasing order of importance):

- [x] Regardless of the `--config-file` arg above, it shows mypy warnings even for files on which we disable mypy, such as `test/test_testing.py`. It seems that VS Code is ignoring the `files` entry in `mypy.ini`.
- [x] It doesn't allow one to get correct warnings for both `mypy.ini` and `mypy-strict.ini` simultaneously (on different files). In fact, one seems to need to reload VS Code after switching which config file the `--config-file` arg in `.vscode/settings.json` points to (although I could be wrong about this).
- [x] It has the potential to get out of sync with the tests in `test/test_type_hints.py`, since those do more than just run `mypy` with a given config file (e.g. they pass other args to `mypy`, just as in the VS Code config above).
- [x] The `--config-file` argument is an absolute path, so the above VS Code config can't be simply copy-pasted between different machines (or checked into version control here).

## Next steps?

It doesn't seem immediately obvious how to solve the first two issues without modifying the VS Code Python extension (or wherever the VS Code mypy integration is implemented). The third one may be a non-issue, though, and the fourth one is just a minor issue of convenience.

Should we include some of the above commands/configs in a wiki somewhere and leave this alone for now, or should we try to address some of these bullet points?

[lint as you type]: https://github.com/pytorch/pytorch/wiki/Lint-as-you-type
[mypy cache]: https://mypy.readthedocs.io/en/stable/command_line.html#incremental-mode
[`torch.random.set_rng_state`]: https://github.com/pytorch/pytorch/blob/5834438090a1b3206347e30968e48f44251a53a1/torch/random.py#L8
[typing pulls]: https://github.com/pytorch/pytorch/pulls?q=is%3Apr+label%3A%22module%3A+typing%22

cc @ezyang @malfet @rgommers @xuzhao9 @gramster

**Comment by ezyang** [comment]
Well, we should fix the cache obliteration problem.

But typically, I just run mypy on *only* the file I'm working on if I need quick feedback. This doesn't pick up typechecking docstrings or stuff like that, but for the code I work on it's not a big deal. Does this work for you?

**Comment by rgommers** [comment]
> We currently have [wiki instructions](https://github.com/pytorch/pytorch/wiki/Lint-as-you-type) for running Flake8 locally in a couple editors, including VS Code. This makes it convenient to correct Flake8 errors before pushing, but we currently don't have similar documentation for running mypy locally.

We do: https://github.com/pytorch/pytorch/wiki/Guide-for-adding-type-annotations-to-PyTorch. Maybe just a matter of making that more visible by linking from `CONTRIBUTING.md`?

> Currently the canonical way to run mypy on PyTorch is via the `test/test_type_hints.py` file; for example:

It's not really, it's simply typing `mypy` in the root of your repo.

**Comment by rgommers** [comment]
Simply using `mypy` also uses the cache correctly. Running `mypy-strict` or doctests only as a check before pushing commits doesn't seem necessary, `mypy` covers pretty much the whole code base and all reverts that have been caused by cross-merge conflicts recently.

**Comment by rgommers** [comment]
I also checked the timings on my machine, which is a 3 year old Intel CPU (12 cores, but `mypy` only uses one):
```
$ rm -rf .mypy_cache/
$ time mypy
Success: no issues found in 1188 source files

real    0m37.743s
user    0m36.821s
sys     0m0.626s
$ time mypy
Success: no issues found in 1188 source files

real    0m0.877s
user    0m0.808s
sys     0m0.067s
```

If it takes 5x longer for you, it's probably the lack of a SSD I'd guess.

**Comment by samestep** [comment]
> Well, we should fix the cache obliteration problem.

@ezyang I should have thought of this yesterday, but this morning I realized that this is actually an easy fix: we can just set [`--cache-dir`](https://mypy.readthedocs.io/en/stable/command_line.html#cmdoption-mypy-cache-dir) to different values for each of the four tests in that suite. I just opened #50539 to address this.

> But typically, I just run mypy on only the file I'm working on if I need quick feedback. This doesn't pick up typechecking docstrings or stuff like that, but for the code I work on it's not a big deal. Does this work for you?

@ezyang That makes sense. The reason I wasn't doing this already is because I was under the mistaken impression that I was supposed to be running `test/test_type_hints.py` (which can't be run on individual files) instead of just `mypy` (which can); see my response to @rgommers below.

> We do: https://github.com/pytorch/pytorch/wiki/Guide-for-adding-type-annotations-to-PyTorch. Maybe just a matter of making that more visible by linking from `CONTRIBUTING.md`?

> It's not really, it's simply typing `mypy` in the root of your repo.

@rgommers Good point, sorry for my unclear wording; what I meant to say was that that wiki page doesn't have any instructions for editor integration like the Flake8 page does. Should I add VS Code instructions to that page, or to the [Lint as you type](https://github.com/pytorch/pytorch/wiki/Lint-as-you-type) page, or neither?

But you are right that that page says to use `mypy` rather than `test/test_type_hints.py`, which is the bit I was missing. I think I overlooked that because

- the page didn't have editor integration instructions, which is what I was really looking for, and
- a coworker specifically told me to run `test/test_type_hints.py`, so I assumed that just running `mypy` wouldn't work.

I do agree that the page should be linked from `CONTRIBUTING.md`, so I just opened #50540 to do that.

> I also checked the timings on my machine, which is a 3 year old Intel CPU (12 cores, but `mypy` only uses one):

> If it takes 5x longer for you, it's probably the lack of a SSD I'd guess.

@rgommers Those timings are surprising, since I am using an SSD (relatively recent 16" MacBook Pro). Are there any other possible causes? I have the repository checked out with all submodules (recursively); does that affect the timing?

**Comment by rgommers** [comment]
> @rgommers Good point, sorry for my unclear wording; what I meant to say was that that wiki page doesn't have any instructions for editor integration like the Flake8 page does. Should I add VS Code instructions to that page, or to the [Lint as you type](https://github.com/pytorch/pytorch/wiki/Lint-as-you-type) page, or neither?

I don't have much of a preference here. There's so many editors that I typically don't add editor-specific info to development guides, but if you think it's useful I'd say please go ahead and add them I'd say. The type annotation page looks like the more appropriate page.

> @rgommers Those timings are surprising, since I am using an SSD (relatively recent 16" MacBook Pro). Are there any other possible causes? I have the repository checked out with all submodules (recursively); does that affect the timing?

Perhaps build differences - I usually build with `USE_DISTRIBUTED=0`, `USE_QNNPACK=0`, `USE_MKLDNN=0`, `USE_FBGEMM=0`, `USE_NNPACK=0`. It would be surprising to me if that would be it, but if it is then maybe it's fixable.

> I do agree that the page should be linked from `CONTRIBUTING.md`, so I just opened #50540 to do that.

Thanks!

**Comment by samestep** [comment]
I'm fairly happy with #50826 (so I'm not planning to work on this any more at the moment once that PR is merged), but it's not completely ideal. Specifically, it still just uses `mypy` instead of [`dmypy`](https://mypy.readthedocs.io/en/stable/mypy_daemon.html), so it loses out on some potential performance improvements. I didn't pursue the daemon route because

- it seems like it'd be more complicated to wrap for our use case, and
- it's currently unsupported by Microsoft's Python extension for VS Code: https://github.com/microsoft/vscode-python/issues/1577
